library/universes.ml

```diff
@@ -756,54 +793,78 @@ let minimize_univ_variables ctx us algs left right cstrs =
 	let lbounds' = 
 	  match compute_lbound (List.map (fun (d,l) -> d, Universe.make l) lower) with
 	  | None -> lbounds
-	  | Some lbound -> LMap.add r (true, false, lbound) lbounds
+	  | Some lbound -> LMap.add r (true, false, lbound, lower_of_list lower)
+                                   lbounds
 	in (Univ.LMap.remove r left, lbounds'))
       left (left, Univ.LMap.empty)
   in
   let rec instance (ctx', us, algs, insts, cstrs as acc) u =
-    let acc, left = 
+    let acc, left, lower =
       try let l = LMap.find u left in
-	    List.fold_left
-	      (fun (acc, left') (d, l) -> 
-	       let acc', (enf,alg,l') = aux acc l in
-	       let l' = 
-		 if enf then Universe.make l 
-		 else l' 
-	       in acc', (d, l') :: left')
-	      (acc, []) l
-      with Not_found -> acc, []
+	  let acc, left, newlow, lower =
+            List.fold_left
+	      (fun (acc, left', newlow, lower') (d, l) ->
+	        let acc', (enf,alg,l',lower) = aux acc l in
+	        let l' =
+		  if enf then Universe.make l
+		  else l'
+	        in acc', (d, l') :: left',
+                   lower_add l d newlow, lower_union lower lower')
+	      (acc, [], LMap.empty, LMap.empty) l
+          in
+          let not_lower (d,l) =
+            Univ.Universe.exists
+              (fun (l,i) ->
+                let d =
+                  if i == 0 then d
+                  else match d with
+                       | Le -> Lt
+                       | d -> d
+                in
+                try let d' = LMap.find l lower in
+                    (* If d is stronger than the already implied lower
+                     * constraints we must keep it *)
+                    compare_constraint_type d d' > 0
```

Comment:
It's unclean I agree, I had to hack around the interface in Univ.Universe. I'm not sure how to help here, I just stared at it long enough. We're checking if (d,l) is already implied by the lower constraints on some level u. If it represents l < u (d is Lt or d is Le and i > 0, the i < 0 case is impossible due to invariants of Univ), and the lower constraints only have l <= u then it is not implied.